### PR TITLE
Warn when converting to Lock subtype

### PR DESCRIPTION
**Pull request by jjonescz**
Test plan: https://github.com/dotnet/roslyn/issues/71888

### CODE REVIEW

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb [hunk 1]
```diff
@@ -845,6 +845,74 @@ End Namespace
...
+        End SyncLock
+
+        Dim l2 As Lock = l1
+        SyncLock l2 ' 1
+        End SyncLock
+
+        Dim l3 As DerivedLock = CType(l2, DerivedLock) ' 2
+        l3 = DirectCast(l2, DerivedLock) ' 3
+        l3 = TryCast(l2, DerivedLock) ' 4
+        SyncLock l3
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:870]: >SyncLock l3 [](http://example.com/codeflow?start=8&length=11)

Do we have a test verifying IL for a scenario like this?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 1), assuming we have or add test mentioned at https://github.com/dotnet/roslyn/pull/72119/files#r1491027442

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb [hunk 2]
```diff
@@ -847,5 +847,139 @@ BC37329: A value of type 'System.Threading.Lock' is not supported in SyncLock. C
...
+    End Class
+
+    Interface IDerivedLock
+    End Interface
+End Namespace
+]]>.Value
+            Dim comp = CreateCompilation(source, options:=TestOptions.ReleaseExe)
+            Dim verifier = CompileAndVerify(comp, expectedOutput:="locked")
+            ' (9, 39) : warning BC42508: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in SyncLock statement.
+            verifier.VerifyDiagnostics(Diagnostic(ERRID.WRN_ConvertingLock, "l2").WithLocation(9, 39))
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:952]: >VerifyDiagnostics(Diagnostic(ERRID.WRN_ConvertingLock, "l2").WithLocation(9, 39)) [](http://example.com/codeflow?start=21&length=81)

Please use `AssertTheseDiagnostics` style helpers to verify diagnostics for VB

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb (hunk 2 above)
Comment (jjonescz) [reply src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:952]: It's not available for the verifier it seems.

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:952]: > It's not available for the verifier it seems.

What prevents us from making it available?

At the very list we should be able to get diagnostics from the verifier and call this helper.

```
    ''' <param name="errors"></param>
    ''' <param name="errs">Expected errors according to this schema
    ''' &lt;expected&gt;[full errors text]&lt;/expected&gt;</param>
    ''' <param name="suppressInfos">True to ignore info-severity diagnostics.</param>
    ''' <remarks></remarks>
    <Extension()>
    Public Sub AssertTheseDiagnostics(errors As ImmutableArray(Of Diagnostic), errs As XElement, Optional suppressInfos As Boolean = True)
        If errs Is Nothing Then
            errs = <errors/>
        End If
        Dim expectedText = FilterString(errs.Value)
        AssertTheseDiagnostics(errors, expectedText, suppressInfos)
    End Sub
```

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

**Pull request merged**